Deprecate {{render helper

josemarluedke opened the pull request:
This implements #13583.

Here is the TODO list from @mixonic.
- [x] Add the deprecation of {{render 'foo'}} in a PR.
- [x]  Add tests for the deprecation of {{render 'foo'}}.
- [x]  Update any tests that still use {{render 'foo'}} to test unrelated behavior to use modern idioms (components).
- [x]  Audit the codebase for any remaining usage, for example in documentation, and refactor that to modern idioms.
- [x]  Audit the guides for any remaining usage of {{render that can be removed.
- [x]  Open a deprecation guide PR on the website.

I'm happy to work in the other items as well, but I'm not sure what to do with the current tests we have. Note that we have quite a few tests around `render into` from a route e.g. https://github.com/emberjs/ember.js/blob/master/packages/ember/tests/routing/basic_test.js#L1813
Also, there are several test failures because of the deprecation. Would love to get some guidance on how to update these tests and migrate the render into tests.

@ErikCH any change you still are interested in taking on some documentation?

josemarluedke committed:
Deprecate {{render}} helper and outlet orphaning

ErikCH commented:
@josemarluedke YES! Absolutely I can look at some documentation when we are ready and get that updated. Thanks for taking this on.

mixonic commented:
@josemarluedke I don't think you should attempt to deprecate `this.render(` on routes as part of this effort. Focusing on `{{render` in isolation would be the happy path.

Thanks for taking this on, please feel free to bug me for reviews etc as you move along :-D

mixonic commented:
Note that you can [ignore deprecations](https://github.com/emberjs/ember.js/blob/aeeddf7c00b42a340494465252481b5414d87009/packages/ember-runtime/tests/controllers/controller_test.js#L118) in tests if there are tests for `{{render` you don't want to change or delete. There should be tests left over for `{{render` in the final version of this PR, but they should be limited to the `{{render` tests and not tests where the use of the deprecated helper is only incidental.

A deprecation guide and audit can totally happen in parallel with this effort, in fact that is even ideal as the two parts of the work can inform each other.

josemarluedke commented:
Yes @mixonic, the goal is just to deprecate `{{render` not `this.render`. Update the title to be more clear. 

I didn't know about the `ignoreDeprecation`, that will definably help to make tests pass.

Should I leave the tests around `render into`?

josemarluedke commented:
This is read for review.

I have added `expectDeprecation` in the tests that uses the render helper and now all the tests are  passing.

I also checked any mentions in the guides for `{{render` and in the ember project itself, no usage of that. 

The deprecation guide is here: emberjs/website#2700. It could be expanded further if necessary.

chadhietala commented:
👍 I had started this in https://github.com/emberjs/ember.js/pull/14246, thanks for carrying it closer to the line. There is code in the router specifically for having a toplevel `{{render}}` that resolves to an `{{outlet}}`. To make sure we don't loose sight of that code when we eventually remove `{{render}}` can you please put a runtime deprecation [here](https://github.com/emberjs/ember.js/blob/9bdb604365a33a67e84364f72bf5c63e7f77a12f/packages/ember-routing/lib/system/router.js#L1314)?

josemarluedke commented:
I have updated per @chadhietala request. Add a new runtime deprecation and updated the deprecation guide in the same PR as before (emberjs/website#2700).

This now closes #14246, #14097 and #13583.

mmun commented:
Nice work!

mixonic commented:
@josemarluedke @chadhietala @ErikCH Thanks very much for this work! ❤️

GavinJoyce commented:
FYI this seems to have introduced some CI test failures:

https://travis-ci.org/emberjs/ember.js/jobs/173976039

```
Test failed: Helpers test: {{render}}:  uses `controller:basic` as the basis for a generated controller when none exists for specified name
    Failed assertion: Died on test #1 	at generateTest (http://localhost:13141/ember.debug.js:43129:19)
	at forEach ([native code])
	at moduleFor (http://localhost:13141/ember.debug.js:43123:33)
	at http://localhost:13141/ember-tests.js:30697:44
	at internalRequire (http://localhost:13141/ember.debug.js:97:21)
	at requireModule (http://localhost:13141/ember.debug.js:44:29)
	at global code (http://localhost:13141/tests/?package=ember-glimmer:190:68): Please refactor `{{render "home"}}` to a component and invoke via `{{home}}`. ('-top-level' @ L1:C0) 
http://localhost:13141/ember.debug.js:5564
Test failed: Helpers test: {{render}}:  generates a controller if none exists
    Failed assertion: Died on test #1 	at generateTest (http://localhost:13141/ember.debug.js:43129:19)
	at forEach ([native code])
	at moduleFor (http://localhost:13141/ember.debug.js:43123:33)
	at http://localhost:13141/ember-tests.js:30697:44
	at internalRequire (http://localhost:13141/ember.debug.js:97:21)
	at requireModule (http://localhost:13141/ember.debug.js:44:29)
	at global code (http://localhost:13141/tests/?package=ember-glimmer:190:68): Please refactor `{{render "home"}}` to a component and invoke via `{{home}}`. ('-top-level' @ L1:C11) 
http://localhost:13141/ember.debug.js:5564
Test failed: Helpers test: {{render}}:  should use controller with the same name as template if present
    Failed assertion: Died on test #1 	at generateTest (http://localhost:13141/ember.debug.js:43129:19)
	at forEach ([native code])
	at moduleFor (http://localhost:13141/ember.debug.js:43123:33)
	at http://localhost:13141/ember-tests.js:30697:44
	at internalRequire (http://localhost:13141/ember.debug.js:97:21)
	at requireModule (http://localhost:13141/ember.debug.js:44:29)
	at global code (http://localhost:13141/tests/?package=ember-glimmer:190:68): Please refactor `{{render "home"}}` to a component and invoke via `{{home}}`. ('-top-level' @ L1:C11) 
http://localhost:13141/ember.debug.js:5564
```

rwjblue commented:
Thanks for fixing those in https://github.com/emberjs/ember.js/issues/14588.  What happened here was that https://github.com/emberjs/ember.js/pull/14582 landed just a few minutes before this PR was merged (and it added these new render tests while fixing a bug). Those new tests didn't have the proper `expectDeprecation` (because the deprecation hadn't landed yet).
